flink-ml-lib/src/main/java/org/apache/flink/ml/recommendation/Swing.java

```diff
@@ -0,0 +1,407 @@
+				if (items.size() < minUserItems || items.size() > maxUserItems) {
+					continue;
+				}
+				for (String item : items) {
+					output.collect(
+						new StreamRecord <>(
+							new Tuple3 <>(
+								user, item, items)));
+				}
+			}
+
+			userItemsMapState.clear();
+		}
+
+		@Override
+		public void processElement(StreamRecord <Tuple2 <String, String>> element) {
+			Tuple2 <String, String> userAndItem =
+				element.getValue();
+			String user = userAndItem.f0;
+			String item = userAndItem.f1;
+			List <String> items = userItemsMap.get(user);
+
+			if (items == null) {
+				ArrayList <String> value = new ArrayList <>();
+				value.add(item);
+				userItemsMap.put(user, value);
+			} else {
+				if (!items.contains(item)) {
+					items.add(item);
+				}
+			}
+		}
+
+		@Override
+		public void initializeState(StateInitializationContext context) throws Exception {
+			super.initializeState(context);
+			userItemsMapState =
+				context.getOperatorStateStore()
+					.getListState(
+						new ListStateDescriptor <>(
+							"userItemsMapState",
+							Types.MAP(
+								Types.STRING,
+								Types.LIST(
+									Types.STRING)))
+					);
+
+			OperatorStateUtils.getUniqueElement(userItemsMapState, "userItemsMapState")
+				.ifPresent(x -> userItemsMap = x);
+
+		}
+
+		@Override
+		public void snapshotState(StateSnapshotContext context) throws Exception {
+			super.snapshotState(context);
+			userItemsMapState.update(Collections.singletonList(userItemsMap));
+		}
+	}
+
+	/**
+	 * Calculate top N similar items of each item.
+	 */
+	private static class CalculateSimilarity
+		extends AbstractStreamOperator <Row>
+		implements OneInputStreamOperator <
+		Tuple3 <String, String, List <String>>,
+		Row>,
+		BoundedOneInput {
+
+		private Map <String, HashSet <String>> userItemsMap = new HashMap <>();
+		private Map <String, HashSet <String>> itemUsersMap = new HashMap <>();
+		private ListState <Map <String, List <String>>> userItemsMapState;
+		private ListState <Map <String, List <String>>> itemUsersMapState;
+
+		final private double alpha = 1.0;
+		final private double userAlpha = 5.0;
+		final private double userBeta = -0.35;
+		final private int topN;
+
+		private CalculateSimilarity(int topN) {this.topN = topN;}
+
+		@Override
+		public void endInput() throws Exception {
+
+			Map <String, Float> userWeights = new HashMap <>(userItemsMap.size());
+			userItemsMap.forEach((k, v) -> {
+				int count = v.size();
+				userWeights.put(k, calculateWeight(count));
+			});
+
+			for (String mainItem : itemUsersMap.keySet()) {
+				List <String> userList = new ArrayList(itemUsersMap.get(mainItem));
+				HashMap <String, Float> id2swing = new HashMap <>();
+
+				for (int i = 0; i < userList.size(); i++) {
+					String u = userList.get(i);
+					for (int j = i + 1; j < userList.size(); j++) {
+						String v = userList.get(j);
+						HashSet <String> interaction = (HashSet <String>) userItemsMap.get(u).clone();
```

Comment:
Thanks for you  suggestion, I checked the code and am sure that a shallow copy can work as intended in this calculation. This copy is used to count number same of items in two `HashSet` instances, and has its own`map` that points to different `HashMap` instance. As for items in the `map`, this method doesn't modify value of items because it is used to count number of `interaction` so I think `clone` can work in this envionment.